daemon/appservices.c

```diff
@@ -0,0 +1,467 @@
+
+    /* Add the current set of applications */
+    applications_t *applications = control_applications(self->asv_control);
+
+    gchar **appnames = stringset_to_strv(applications_available(applications));
+
+    for( guint i = 0; appnames[i]; ++i ) {
+        char *appname = appnames[i];
+
+        appinfo_t *appinfo = applications_appinfo(applications, appname);
+
+        if( appinfo && appinfo_dbus_auto_start(appinfo) ) {
+            stringset_remove_item(appnames_to_remove, appname);
+
+            appservices_write_service_file(self, appname, appinfo);
+        }
+    }
+
+    g_strfreev(appnames);
+
+    /* Remove any service files that weren't matched to an application. */
+    appnames = stringset_to_strv(appnames_to_remove);
+    for( guint i = 0; appnames[i]; ++i ) {
+        appservices_application_removed(self, appnames[i]);
+    }
+    g_strfreev(appnames);
+    stringset_delete(appnames_to_remove);
+}
+
+void
+appservices_application_changed(appservices_t *self, const char *appname, appinfo_t *appinfo)
+{
+    if( appinfo_dbus_auto_start(appinfo) ) {
+        appservices_write_service_file(self, appname, appinfo);
+    }
+    else {
+        appservices_application_removed(self, appname);
+    }
+}
+
+void
+appservices_application_added(appservices_t *self, const char *appname, appinfo_t *appinfo)
+{
+    if( appinfo_dbus_auto_start(appinfo)) {
+        appservices_write_service_file(self, appname, appinfo);
+    }
+}
+
+void
+appservices_application_removed(appservices_t *self, const char *appname)
+{
+    if( !self->asv_run_dir ) {
+        return;
+    }
+
+    const serviceinfo_t *service = g_hash_table_lookup(self->asv_service_lut, appname);
+
+    if( service) {
+        char *service_filename = appservices_service_filename(self, service->name);
+
+        log_info("appservices(%s) remove service file %s", appname, service_filename);
+
+        unlink(service_filename);
+        g_free(service_filename);
+
+        g_hash_table_remove(self->asv_service_lut, appname);
+
+        control_on_appservices_change(self->asv_control);
+    }
+}
+
+void
+appservices_update_user(appservices_t *self)
+{
+    uid_t uid = control_current_user(self->asv_control);
+
+    if( self->asv_uid != uid ) {
+        self->asv_uid = uid;
+
+        g_free(self->asv_run_dir);
+        self->asv_run_dir = NULL;
+
+        /* Get the gid and run directory of the current user */
+        if( self->asv_uid != SESSION_UID_UNDEFINED ) {
+            long initlen = sysconf(_SC_GETPW_R_SIZE_MAX);
+            size_t len = 131072; // ARG_MAX;
+            if( initlen != -1 ) {
+                len = (size_t) initlen;
+            }
+
+            struct passwd result;
+            struct passwd *resultp;
+            char *buffer = malloc(len);
+
+            if( !getpwuid_r(self->asv_uid, &result, buffer, len, &resultp) ) {
+                self->asv_gid = result.pw_gid;
+                self->asv_run_dir = g_strdup_printf("/run/user/%lld",
```

Comment:
It is not likely that `/run/user` changes at any point but wouldn't it be good to have this as non-hardcoded path?

Reply:
Probably, the only reason I didn't to begin with was that I followed the lead of prompter.c when it creates the path for the session dbus path.